flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java

```diff
@@ -517,6 +517,21 @@ public void checkpointStopped(long checkpointId) {
 		}
 	}
 
+	@Override
+	public void convertToPriorityEvent(int sequenceNumber) throws IOException {
+		boolean firstPriorityEvent;
+		synchronized (receivedBuffers) {
+			checkState(!channelStatePersister.hasBarrierReceived());
+			SequenceBuffer toPrioritize = receivedBuffers.getAndRemove(
+				sequenceBuffer -> sequenceBuffer.sequenceNumber == sequenceNumber);
+			checkState(!toPrioritize.buffer.isBuffer());
+			firstPriorityEvent = addPriorityBuffer(toPrioritize);
+		}
+		if (firstPriorityEvent) {
+			notifyPriorityEvent(sequenceNumber);
+		}
+	}
+
```

Comment:
This is different to what we discussed offline @AHeise . Instead of switching to the pattern of keeping the `CheckpointBarrier` at the tail of the input queue, and spilling everything that falls between `CheckpointBarrier` and it's announcement, I decided to keep for the sake of simplicity the current approach of using priority events. And instead I'm just prioritizing the previously announced aligned `CheckpointBarrier`.

Reply:
@pnowojski , I thought that `firstPriorityEvent` variable is unused and could be removed like following:
```
public void convertToPriorityEvent(int sequenceNumber) throws IOException {
		synchronized (receivedBuffers) {
			checkState(!channelStatePersister.hasBarrierReceived());
			SequenceBuffer toPrioritize = receivedBuffers.getAndRemove(
				sequenceBuffer -> sequenceBuffer.sequenceNumber == sequenceNumber);
			checkState(!toPrioritize.buffer.isBuffer());
            if (addPriorityBuffer(toPrioritize)) {
			   notifyPriorityEvent(sequenceNumber);
	 	    }
		}
	}
```

Reply:
We tend to avoid firing notifications under locks if possible. We don't know what code will be executed, and that code can try to acquire some higher level lock (for example `inputChannelsWithData` or `requestLock` from `SingleInputGate`) compared to `receivedBuffers`. If we then have another code path acquiring those locks in opposite order we can have a traditional example of the deadlock.

I'm not sure if this is still an issue here (definitely there was a problem with notifications), but it's nevertheless our best practise to invoke callbacks/notifications/complete futures outside of synchronized sections if possible.